packages/type-utils/tests/requiresQuoting.test.ts

```diff
@@ -0,0 +1,84 @@
+import { requiresQuoting } from '../src';
+
+describe('getDeclaration', () => {
+  describe('valid identifier', () => {
+    it('upper and lower case alphabet', () => {
+      const name = 'fooBar';
+      const result = requiresQuoting(name);
+      expect(result).toBe(false);
+    });
+
+    it('upper and lower case alphabet with number not first', () => {
+      const name = 'foo1234Bar';
+      const result = requiresQuoting(name);
+      expect(result).toBe(false);
+    });
+
+    it('start with dollorSign', () => {
+      const name = '$bar';
+      const result = requiresQuoting(name);
+      expect(result).toBe(true);
+    });
+    it('include dollorSign not start position', () => {
+      const name = 'foo$bar';
+      const result = requiresQuoting(name);
+      expect(result).toBe(true);
+    });
+
+    it('start with underScore', () => {
+      const name = '_bar';
+      const result = requiresQuoting(name);
+      expect(result).toBe(true);
+    });
+    it('include underScore not start position', () => {
+      const name = 'foo_bar';
+      const result = requiresQuoting(name);
+      expect(result).toBe(true);
+    });
+
+    it('non-alphabet characters', () => {
+      const name = 'café';
+      const result = requiresQuoting(name);
+      expect(result).toBe(false);
+    });
+  });
+
+  describe('invalid identifier', () => {
+    it('only number', () => {
+      const name = '12345';
+      const result = requiresQuoting(name);
+      expect(result).toBe(true);
```

Comment:
I think this test case return true is weird. 

because I don't think quotation marks are necessary if the name of the property variable consists entirely of numbers.

If the function is incorrect in returning true for this test case, we will create a separate issue and remove the test case.

Reply:
Ooh, good eyes 👀. Yeah I guess it depends how `requiresQuoting` is intended to be used.... 

You can _define_ a property starting with a number, but you can't _access_ it that way using dot notation, whereas you can with the other special characters here. 
```ts
const objectWithCrapPropertyNames = {
  12345: "legal syntax!",
  $foo: "foo",
  _bar: "bar",
}

// syntax error
objectWithCrapPropertyNames.12345
// works
objectWithCrapPropertyNames.$foo
// works
objectWithCrapPropertyNames._bar
```

This was extracted to a util to support the naming-convention requiresQuotes option (https://github.com/typescript-eslint/typescript-eslint/pull/2813)... but existed before that as a utility in switch-exhaustiveness check for autofixing with code that performs member access, in which case the quoting _is_ required on the number (https://github.com/typescript-eslint/typescript-eslint/pull/2207).

That it can handle numbers in this way is actually explicitly tested for:
https://github.com/typescript-eslint/typescript-eslint/blob/82cb9dd580f62644ed988fd2bf27f519177a60bd/packages/eslint-plugin/tests/rules/switch-exhaustiveness-check.test.ts#L1861-L1897

So I'd say it probably needs to work the way it currently does for its intended usages to work - but maybe you can leave a comment to clarify that, either on the test cases or the function itself?

Reply:
Thank you for telling us about the history. I completely understand why the test case is needed.

Regarding comments, I added them because I thought it would be good to have appropriate comments for both functions and test code. It would be a good idea to check them out.